ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -0,0 +1,225 @@
+  private QueryErrorClassifier errorClassifier ;
+  private KafkaStreamsNamedTopologyWrapper kafkaStreams;
+  private ImmutableMap<String, Object> streamsProperties;
+  private final QueryMetadataImpl.TimeBoundedQueue queryErrors;
+  private final Map<String, PersistentQueriesInSharedRuntimesImpl> metadata;
+  public final Map<QueryId, Set<SourceName>> sources;
+
+  public SharedKafkaStreamsRuntimeImpl(final KafkaStreamsBuilder kafkaStreamsBuilder,
+                                       final int maxQueryErrorsQueueSize,
+                                       final Map<String, Object> streamsProperties) {
+    this.kafkaStreamsBuilder = kafkaStreamsBuilder;
+    kafkaStreams = kafkaStreamsBuilder.build(streamsProperties);
+    queryErrors
+        = new QueryMetadataImpl.TimeBoundedQueue(Duration.ofHours(1), maxQueryErrorsQueueSize);
+    this.streamsProperties = ImmutableMap.copyOf(streamsProperties);
+    metadata = new ConcurrentHashMap<>();
+    sources = new ConcurrentHashMap<>();
+    kafkaStreams.setUncaughtExceptionHandler(this::uncaughtHandler);
+    kafkaStreams.start();
+  }
+
+  public void markSources(final QueryId queryId, final Set<SourceName> sourceNames) {
+    sources.put(queryId, sourceNames);
+  }
+
+  public void addQuery(
+          final QueryErrorClassifier errorClassifier,
+          final Map<String, Object> streamsProperties,
+          final PersistentQueriesInSharedRuntimesImpl persistentQueriesInSharedRuntimesImpl,
+          final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    if (metadata.containsKey(queryId.toString())) {
+      kafkaStreams.removeNamedTopology(queryId.toString());
+    }
+    metadata.put(queryId.toString(), persistentQueriesInSharedRuntimesImpl);
+    LOG.debug("mapping {}", metadata);
+  }
+
+  public StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+  ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (PersistentQueriesInSharedRuntimesImpl query: metadata.values()) {
+        query.getListener().onError(metadata.get(query.getQueryId().toString()), queryError);
```

Comment:
Related to my other comment, this is the part I was concerning about: if it throws we just mark all runtimes / queries as failed --- even if some queries are on different runtimes. Is that okay?

Reply:
We decided that it was okay for a first round. It will require some more changes to make it possible to distinguish which `NamedTopology` the error comes from. But it shouldn't stop the query it will just let the metrics know there has been an error